### PR TITLE
fix: logger formatter

### PR DESCRIPTION
**Pull request by DDDHuang**

### CODE REVIEW

**Comment by DDDHuang** [comment]
TODO: more date funs & move date func in emqx_calendar.erl

File: etc/emqx.conf [hunk 1]
```diff
@@ -458,7 +458,35 @@ log.file = emqx.log
 
 ## Log formatter
 ## Value: text | json
-#log.formatter = text
+log.formatter = text
+
+## Text logger, date format.
+##
+## Value: rfc3339 | FormatString
+##   FormatString:
```
Comment (terry-xiaoyu) [inline etc/emqx.conf:466]:
```
## Value: rfc3339 | FORMAT
##     Where FORMAT is the format string of the timestamp. Supported specifiers:
```

File: etc/emqx.conf [hunk 2]
```diff
@@ -458,7 +458,35 @@ log.file = emqx.log
...
+##     %Y: year
+##     %m: month
+##     %d: day
+##     %H: hour
+##     %M: minute
+##     %S: second
+##     %6N: microseconds
+##     %3N: milliseconds
+##     %z: timezone, [+|-]HH:MM or [+|-]HH:MM:SS. Depends on `log.formatter.text.date.timezone`
```
Comment (terry-xiaoyu) [inline etc/emqx.conf:475]: We can support `%z`, `%:z`,  `%::z` just like what the GNU `date` command supports.

File: etc/emqx.conf [hunk 3]
```diff
@@ -458,7 +458,35 @@ log.file = emqx.log
...
+## Value: rfc3339 | FormatString
+##   FormatString:
+##     %Y: year
+##     %m: month
+##     %d: day
+##     %H: hour
+##     %M: minute
+##     %S: second
+##     %6N: microseconds
+##     %3N: milliseconds
```
Comment (terry-xiaoyu) [inline etc/emqx.conf:474]: Add nano sec here?
```
%N   nanoseconds (000000000..999999999)
```

File: etc/emqx.conf [hunk 4]
```diff
@@ -458,7 +458,35 @@ log.file = emqx.log
...
+##     %z: timezone, [+|-]HH:MM or [+|-]HH:MM:SS. Depends on `log.formatter.text.date.timezone`
+## For example:
+## log.formatter.text.date.format = emqx-server-date: %Y-%m-%dT%H:%M:%S.%6N %z
+## Default: rfc3339
+# log.formatter.text.date.format = rfc3339
```
Comment (terry-xiaoyu) [inline etc/emqx.conf:479]:
```suggestion
## For example:
## log.formatter.text.date.format = %Y-%m-%d %H:%M:%S.%3N %:z
##
## Default: rfc3339
# log.formatter.text.format = rfc3339
```

File: etc/emqx.conf [hunk 5]
```diff
@@ -458,7 +458,35 @@ log.file = emqx.log
...
+
+## Text logger, timezone.
+## Only takes effect when log.formatter.text.date.format not rfc3339.
+##
+## Value: local | z | UTC TimeOffset
+## TimeOffset: [+|-]HH:MM or [+|-]HH:MM:SS, for example: +08:00.
+## local is the system local time timezone.
+## z: UTC zulu timezone.
+## Default: local
+# log.formatter.text.date.timezone = local
```
Comment (terry-xiaoyu) [inline etc/emqx.conf:489]:
```suggestion
## The time_offset for the timestamp printed by the logger.
##
## Value: local | z | TimeOffset
## - local is the system local time timezone.
## - z: Universal Coordinated Time (UTC).
## - TimeOffset: [+|-]HH:MM or [+|-]HH:MM:SS, for example: +08:00.
##
## Examples:
##   log.formatter.time_offset = z
##   log.formatter.time_offset = +08:00
##
## Default: local
# log.formatter.time_offset = local
```

Comment (DDDHuang) [reply etc/emqx.conf]: removed

File: src/emqx_calendar.erl [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+            case lists:member(Key, ?DATE_ZONE_NAME) of
+                true ->
+                    do_parse_date_str(Tail, Formatter, Result#{parsed_offset => offset_second(DatePart)});
+                false ->
+                    do_parse_date_str(Tail, Formatter, Result)
+            end
+    end.
+
+date_size(Str) when is_list(Str) -> erlang:length(Str);
+date_size(DateName) ->
```
Comment (zmstone) [inline src/emqx_calendar.erl:411]: nit: use function clauses

File: src/emqx_calendar.erl [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+        millisecond => 3,
+        microsecond => 6,
+        nanosecond => 9,
+        timezone => 5,
+        timezone1 => 6,
+        timezone2 => 9
+    },
+    maps:get(DateName, Map).
+
+dm(Month) ->
```
Comment (zmstone) [inline src/emqx_calendar.erl:428]: nit: use function clauses

File: etc/emqx.conf [hunk 8]
```diff
@@ -458,7 +458,29 @@ log.file = emqx.log
...
+##     %S: second
+##     %N: nanoseconds (000000000 - 999999999)
+##     %6N: microseconds (00000 - 999999)
+##     %3N: milliseconds (000 - 999)
+##     %z: timezone, [+-]HHMM
+##     %:z: timezone, [+-]HH:MM
+##     %::z: timezone, [+-]HH:MM:SS
+##
+## For example:
+## log.formatter.text.date.format = %Y-%m-%dT%H:%M:%S.%6N %:z
```
Comment (terry-xiaoyu) [inline etc/emqx.conf:481]: For example:
  log.formatter.text.date.format = %Y-%m-%dT%H:%M:%S.%6N %:z
  log.formatter.text.date.format = %Y-%m-%d %H:%M:%S.%3N

File: src/emqx_calendar.erl [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+                end;
+            (nanosecond, V, Res) ->
+                Res + V;
+            (parsed_offset, V, Res) ->
+                Res - V
+        end,
+    Count = maps:fold(Counter, 0, DateInfo) - (?SECONDS_PER_DAY * Precise),
+    erlang:convert_time_unit(Count, PrecisionUnit, Unit).
+
+precise(#{nanosecond := _}) -> {1000_000_000, nanosecond};
```
Comment (zmstone) [inline src/emqx_calendar.erl:387]:
```suggestion
precision(#{nanosecond := _}) -> {1000_000_000, nanosecond};
```

File: src/emqx_calendar.erl [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+    UniversalSecs = calendar:datetime_to_gregorian_seconds(UniversalTime),
+    LocalSecs - UniversalSecs;
+do_offset_second(Offset) when is_binary(Offset) ->
+    do_offset_second(erlang:binary_to_list(Offset));
+do_offset_second("Z") -> 0;
+do_offset_second("z") -> 0;
+do_offset_second(Offset) when is_list(Offset) ->
+    Sign = hd(Offset),
+    ((Sign == $+) orelse (Sign == $-))
+        orelse error({bad_zone, Offset}),
```
Comment (terry-xiaoyu) [inline src/emqx_calendar.erl:125]: bad_time_offset

File: src/emqx_calendar.erl [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+    {HourStr, MinuteStr, SecondStr} =
+        case string:tokens(HM, ":") of
+            [H, M] ->
+                {H, M, "0"};
+            [H, M, S] ->
+                {H, M, S};
+            [HHMM] when erlang:length(HHMM) == 4 ->
+                {string:sub_string(HHMM, 1,2), string:sub_string(HHMM, 3,4), "0"};
+            _ ->
+                error({bad_zone, Offset})
```
Comment (terry-xiaoyu) [inline src/emqx_calendar.erl:138]: bad_time_offset

File: src/emqx_logger_textfmt.erl [hunk 12]
```diff
@@ -28,11 +28,28 @@
         , gl        % not interesting
         ]).
 
-check_config(X) -> logger_formatter:check_config(X).
+check_config(Config0) ->
+    Config = maps:without([timezone_offset, timezone, date_format], Config0),
```
Comment (terry-xiaoyu) [inline src/emqx_logger_textfmt.erl:32]: Config = maps:without([date_format], Config0),

File: src/emqx_logger_textfmt.erl [hunk 13]
```diff
@@ -28,11 +28,28 @@
...
+    logger_formatter:check_config(Config).
 
-format(#{msg := Msg0, meta := Meta} = Event, Config) ->
+format(#{msg := Msg0, meta := Meta} = Event,
+       #{date_format := rfc3339, template := Template0} = Config) ->
     Msg = maybe_merge(Msg0, Meta),
-    logger_formatter:format(Event#{msg := Msg}, Config).
+    Template = [time | Template0],
+    logger_formatter:format(Event#{msg := Msg}, Config#{template => Template});
+
```
Comment (zmstone) [inline src/emqx_logger_textfmt.erl:40]:
```suggestion
```

File: src/emqx_calendar.erl [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+    , timezone2
+    ]).
+
+formatter(FormatterStr) when is_list(FormatterStr) ->
+    formatter(list_to_binary(FormatterStr));
+formatter(FormatterBin) when is_binary(FormatterBin) ->
+    do_formatter(FormatterBin, []).
+
+offset_second(Offset) ->
+    do_offset_second(Offset).
```
Comment (zmstone) [inline src/emqx_calendar.erl:58]: nit: no need for the `do_` version implementation ?

**Review by terry-xiaoyu — APPROVED** [review]

**Pull request merged**